larpandoracontent/LArHelpers/LArHierarchyHelper.cc

```diff
@@ -0,0 +1,1171 @@
+/**
+ *  @file   larpandoracontent/LArHelpers/LArHierarchyHelper.cc
+ *
+ *  @brief  Implementation of the lar hierarchy helper class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/PdgTable.h"
+#include "Pandora/StatusCodes.h"
+
+#include "larpandoracontent/LArHelpers/LArHierarchyHelper.h"
+
+namespace lar_content
+{
+
+using namespace pandora;
+
+LArHierarchyHelper::MCHierarchy::MCHierarchy(const ReconstructabilityCriteria &recoCriteria) :
+    m_recoCriteria(recoCriteria),
+    m_pNeutrino{nullptr}
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+LArHierarchyHelper::MCHierarchy::~MCHierarchy()
+{
+    for (const Node *pNode : m_rootNodes)
+        delete pNode;
+    m_rootNodes.clear();
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void LArHierarchyHelper::MCHierarchy::FillHierarchy(const MCParticleList &mcParticleList, const CaloHitList &caloHitList,
+    const bool foldToPrimaries, const bool foldToLeadingShowers)
+{
+    auto predicate = [](const MCParticle *pMC) { return std::abs(pMC->GetParticleId()) == NEUTRON; };
+    m_mcToHitsMap.clear();
+    for (const CaloHit *pCaloHit : caloHitList)
+    {
+        try
+        {
+            const MCParticle *pMC{MCParticleHelper::GetMainMCParticle(pCaloHit)};
+            m_mcToHitsMap[pMC].emplace_back(pCaloHit);
+        }
+        catch(const StatusCodeException&)
+        {
+        }
+    }
+
+    if (foldToPrimaries && !foldToLeadingShowers)
+    {
+        MCParticleSet primarySet;
+        m_pNeutrino = LArHierarchyHelper::GetMCPrimaries(mcParticleList, primarySet);
+        MCParticleList primaries(primarySet.begin(), primarySet.end());
+        if (m_recoCriteria.m_removeNeutrons)
+            primaries.erase(std::remove_if(primaries.begin(), primaries.end(), predicate), primaries.end());
+        for (const MCParticle *pPrimary : primaries)
+        {
+            MCParticleList allParticles{pPrimary};
+            if (!m_recoCriteria.m_removeNeutrons)
+            {
+                LArMCParticleHelper::GetAllDescendentMCParticles(pPrimary, allParticles);
+            }
+            else
+            {   // Collect track-like and shower-like particles together, but throw out neutrons and descendents
+                MCParticleList neutrons;
+                LArMCParticleHelper::GetAllDescendentMCParticles(pPrimary, allParticles, allParticles, neutrons);
```

Comment:
The neutrons here are just thrown away - later on, an analogous list is labelled as "dummy"; could perhaps make a consistent choice here... (I'd previously put in a similar comment, then cancelled it as the logic wasn't fully clear to me yet. sorry)

Reply:
Yes, originally I was always keeping neutrons and just tagging them separately, so this is a legacy of providing an option to throw them away altogether. I'll alter the cases where there is no folding that might still keep them as leading particles to use 'dummy' everywhere. I think Leigh made a similar comment as well, but I don't see it anywhere.

Reply:
Ah, I removed it as it was in the same comment as something that ended up untrue.